### ISSUE TITLE
Logging

### ISSUE THREAD

**Issue by tgp1994** [wokhansoft/WFN#20]
Currently, logs are saved in the base directory of the application. The purpose of this issue is to establish a plan for where logs will be stored. Ultimately I imagine we can make this a user-selectable option, but by default it could go like this:

1. Pick Windows Event Reporting by default
2. Save the log file to a temp folder (fall back here if writing to WER fails)
3. Save the log file to the current working directory
3. Don't do any writing at all (unwritable file system?)

## Prerequisites
The discussion of logging was brought about from discussing the need for a portable mode or option where the application can run self-contained in its own directory (this is how it currently operates), rather than being installed to the system. In portable mode, it may make sense to change the priorities of logging.

### To log or not to log?
As WFN becomes more developed, it may be worthwhile to ask if logging should be enabled by default at all. In a production environment, it may be desirable to have as little logging as possible, at least having some sort of management system in place to prevent log files from growing out of control. Logging could be disabled by default in release mode (but optionally enabled by the user of course), then enabled by default in debug mode. In addition, WFN could keep a memory buffer of events, then dump them to the requested log location when necessary.

**Comment by harrwiss** [comment]
Since we have log4net the location of the file can be changed in the *WFN.config.log4net* config file or turned off completely. It would also be possible to send the log to a log-server with another appender.

One issue we should look at is, that people tend to install a program in the *Program Files* folder which has some restrictions when writing logs (I know this was an issue before log4net).

**Comment by wokhan** [comment]
Exactly my thoughts (hence the "in progress" status and the assignation to your side ;-) ).
On a side note, people might be against it, but telemetry is also sometimes really useful... Especially when users give us a unique ID which help us find out what happened on a centralized logs server.
Could be another option (either replacing - or along with - the file/eventlog/whatever logs).

**Comment by harrwiss** [comment]
Unfortunately the logging is not working at all anymore - probably a problem happened during merge in the build config. Need to investigate.

**Comment by wokhan** [comment]
Maybe I broke something when changing the output folder, if the log4net config file isn't copied to the output anymore...
You'd have to select it and update the "copy to output dir" option again... Maybe.

**Comment by harrwiss** [comment]
Thanks for the hint - that was also a problem 😃 

The other problem was, that in the assemblyinfo.cs the following line was used to find the configuration:
>// log4net - look for a configuration file in the base dir
[assembly: log4net.Config.XmlConfigurator(ConfigFile="WFN.config.log4net", Watch = true)]

Alas! took me a while to figure out that it's not used anymore in nc31 (I will remove the file now).  I have a workaround which is implemented in LogHelper now (maybe better anyway to avoid such problems).

**Comment by wokhan** [comment]
Yes sorry I also missed that! At least... thanks to me forgetting some things in the process, you're now almost a .net core migration expert 😉

**Comment by harrwiss** [comment]
No problem, I also did not remember that I've put it in there 🤕
